src/v/cluster/archival_metadata_stm.cc

```diff
@@ -1311,6 +1265,7 @@ void archival_metadata_stm::apply_add_segment(const segment& segment) {
           "Can't add segment: {}, previous segment: {}",
           meta,
           last);
+        maybe_notify_waiter(errc::inconsistent_stm_update);
```

Comment:
Shouldn't we throw here and let the try-catch bock in apply propagate the error?

Reply:
it will be registered and propagated as an exception and we need an error code
this can be done of cause but will require more code and separate code path to call `maybe_notify_waiter` with the error code

Reply:
Can't then we have a case where the STM is a bit behind, we call replicate (we even include the fence command) and set `_active_operation_res` but then ... `maybe_notify_waiter` is called for an older command? What's preventing this?

Reply:
the protocol is to call `sync` first and only after that call `replicate`
so basically, you're calling `sync`, getting read-write-fence value and then replicating data

Reply:
do you think some external failsafe mechanism should be added to prevent incorrect usage?

Reply:
Yeah, I believe this implementation is still problematic. Anything that prevents the caller to call sync then do some other async thing and then call replicate? The correctness of the STM still relies on assumptions which is not better than what we had from what I understand.

Reply:
I think I solved this problem by forcing `sync` in case if the STM is catching up. So we're only setting the promise when the STM is up to date and we're holding the lock so it's guaranteed that the promise will get the correct result.